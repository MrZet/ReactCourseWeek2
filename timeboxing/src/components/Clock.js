import React from 'react'

const addLeadingZeroes = (number, places = 2) => number.toString().padStart(places,'0');
        
function validateDatePart(datePart, maxValue)
{
    if(datePart<0)
        return 0;
    else if(datePart>maxValue)
        return maxValue;
    else 
        return datePart;
}

function formatDate(datePart, maxValue, places)
{
    datePart = datePart.toString();
    datePart = validateDatePart(datePart, maxValue);
    datePart = addLeadingZeroes(datePart, places)

    return datePart;
}

function Clock({hours=0, minutes=0, seconds=0, miliseconds=0, className="", children})
{
    hours = formatDate(hours,23);
    minutes = formatDate(minutes,59);
    seconds = formatDate(seconds,59);
    miliseconds = formatDate(miliseconds,999,3);

    return (<h2 className = {"Clock "+className}>{children} {hours}:{minutes}:{seconds}:{miliseconds}</h2>);
};

export default Clock;