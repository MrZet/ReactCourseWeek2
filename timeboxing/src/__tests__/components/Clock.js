import React from 'react'
import Clock from '../../components/Clock'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

var root = null;
var clockRenderer = null;

describe('<Clock/>', () => {
    describe('When given time parts (DOM)',()=>{
        beforeEach(()=>{
            root = document.createElement("div");
            ReactDOM.render(<Clock hours={12} minutes={13} />, root);
        })
        it('render and put H2 element', () => {
            expect(root.childNodes[0].nodeName).toEqual('H2')
        });   
        it('sets a Clock class name', () => {
            expect(root.childNodes[0].className).toMatch(/Clock/)
        });   
        it('renders time properly', () => {
            expect(root.childNodes[0].textContent).toMatch(/12:13/)
        });   
    });

    describe('When given time parts (REACT-TEST-RENDERER)',()=>{
        beforeEach(()=>{
            clockRenderer = renderer.create(
                <Clock hours={12} minutes={13} />
            );
        })
        it('render properly', () => {
            expect(clockRenderer.toJSON()).toMatchSnapshot();
        });
        it('render and put H2 element', () => {
            expect(root.childNodes[0].nodeName).toEqual('H2')
        });   
        it('sets a Clock class name', () => {
            expect(clockRenderer.toJSON().props).toMatchObject({"className": expect.stringMatching(/Clock/)});
        });   
        it('renders time properly', () => {
            expect(clockRenderer.toJSON().children).toEqual(expect.arrayContaining(["12","13"]))
        });   
    });
})

