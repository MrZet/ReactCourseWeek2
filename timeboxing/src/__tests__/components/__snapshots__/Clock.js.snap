// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock/> When given time parts (REACT-TEST-RENDERER) render properly 1`] = `
<h2
  className="Clock "
>
   
  12
  :
  13
  :
  00
  :
  000
</h2>
`;
